proxy-server/go/proxy-main.go

```diff
@@ -0,0 +1,306 @@
+package main
+
+import (
+	"crypto/rand"
+	"fmt"
+	"github.com/dgrijalva/jwt-go"
+	"log"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+)
+
+//SECURITY NOTICE: Some older versions of Go have a security issue in the cryotp/elliptic.
+//Recommendation is to upgrade to at least 1.8.3.
+
+var defaultURL = "http://localhost"           // resource default url. port 80 of the localhost
+var fallbackURL = "http://localhost/fallback" // not authorized fallback
+var jwtKey []byte
+var sharedKey []byte
+
+var minutesExpire = 60 //  1 hour to expiration
+var debug = true
+var defaultPort = "8080"
+var mutex sync.Mutex
+var tokenMap = make(map[string]string)
+
+const keyLength = 512 / 8
+
+const (
+	errGeneric          = iota
+	errNoCookie         = iota
+	errBadSignature     = iota
+	errExpired          = iota
+	errTokenInvalid     = iota
+	errSignatureInvalid = iota
+)
+
+// Get env var or default
+func getEnv(key, dflt string) string {
+	value, ok := os.LookupEnv(key)
+	if ok {
+		return value
+	}
+	return dflt
+}
+
+func generateCookieSigningKey(n int) ([]byte, error) {
+	const letters = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz-"
+	b := make([]byte, n)
+	_, err := rand.Read(b)
+	if err != nil {
+		return nil, err
+	}
+	for i, b1 := range b {
+		b[i] = letters[b1%byte(len(letters))]
+	}
+	return b, nil
+}
+
+// Serve a reverse proxy for a given url
+func serveReverseProxy(targetURL string, res http.ResponseWriter, req *http.Request) {
+	url, _ := url.Parse(targetURL)
+	proxy := httputil.NewSingleHostReverseProxy(url)
+
+	// Modify the headers for  SSL redirection
+	req.URL.Host = url.Host
+	req.URL.Scheme = url.Scheme
+	req.Header.Set("X-Forwarded-Host", req.Header.Get("Host"))
+	req.Host = url.Host
+	proxy.ServeHTTP(res, req) // non blocking
+}
+
+type extendedClaims struct {
+	resourceAddress string `json:"resource"`
+	//	reqToken string `json:"req_token"`
+	jwt.StandardClaims
+}
+
+func setNewCookie(res http.ResponseWriter, req *http.Request) {
+
+	claims := &extendedClaims{}
+	expirationTime := time.Now().Add(time.Duration(minutesExpire) * time.Minute)
+	claims.ExpiresAt = expirationTime.Unix()
+	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
+	tokenString, err := token.SignedString(jwtKey)
+	if err != nil {
+		res.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// Set the new token as the users `saturn_token` cookie
+	http.SetCookie(res, &http.Cookie{
+		Name:    "saturn_token",
+		Value:   tokenString,
+		Expires: expirationTime,
+	})
+}
+
+// The default fallback handler.
+func handleRequestFallBack(res http.ResponseWriter, req *http.Request) {
+	res.WriteHeader(http.StatusUnauthorized)
+	//		res.WriteHeader(http.StatusBadRequest)
+	fmt.Fprintf(res, "\nWe are sorry, access denied.")
```

Comment:
This should get sent in the response, not the log.

Reply:
This _is_ a response ! :)  (Writer)

Reply:
Oh! Yes it is :)